### ISSUE TITLE
Enhancement: move the test_JSON tree

### ISSUE THREAD

**Issue by lcn2** [ioccc-src/mkiocccentry#449]
## Background

> ...shouldn't `vermod.sh` be put under `jparse` and not `soup` ?

The propose of `vermod.sh` is _IOCCC specific_: to update version numbers that matter to the _mkiocccentry repo_.  

While `jparse/test_jparse/test_JSON/` contains what seems to be _IOCCC data_, data that `vermod.sh` is designed to modify: what the _JSON parser_ needs are JSON test files.  It just happens that the _IOCCC inspired_  `jparse/test_jparse/test_JSON/` happened to "_foot the bill_" (as that expression goes).

## move the test_JSON tree

We suggest that `jparse/test_jparse/test_JSON/` should be moved to `test_ioccc/test_JSON/` and that `vermod.sh` be modified to impact that directory tree.

Now, the _JSON parser_ needs test JSON files of both valid and invalid JSON variety.  So while the **CONTENT** under `jparse/test_jparse/test_JSON/` should be moved to `test_ioccc/test_JSON/`, the _JSON parser_ **STILL NEEDS** a `parse/test_jparse/test_JSON/` test tree: just something that is **NOT** _IOCCC / mkiocccentry_ related.  We suggest that the "something" that the _JSON parser_ should do is to make the new `jparse/test_jparse/test_JSON/` based on the theme that `jparse/test_jparse/sample_JSON/party.json` uses.

For the above to be accomplished well, the respective tools that test the JSON parser need a clear mechanism to identify the directory under which the _test_JSON_ stuff is found.  Moreover, the directory structure and filename structure under the current `jparse/test_jparse/test_JSON/` directory needs to be maintained.  

The current `jparse/test_jparse/jparse_test.sh` does this with the `-d json_tree` argument that defaults to the "`$JSON_TREE`" (`./jparse/test_jparse/test_JSON`) value.   So there may not be much that needs to change.  Perhaps what is needed (in addition to the above melted tree move and new _test_JSON_ tree contents based on _party.json_ theme) is that the _test_ Makefile rule of `json/test_jparse/Makefile` needs to point at the _test_JSON_ tree contents based on the _party.json_ theme WHEREAS the  _test_ Makefile rule of `test_ioccc/Makefile` would invoke `jparse/test_jparse/jparse_test.sh` executed with `-d test_ioccc/test_JSON` (containing the current JSON contents that is under `jparse/test_jparse/test_JSON/` now).

IMPORTANT NOTE: The new contents of `jparse/test_jparse/test_JSON/` are **NOT** needed right now.  Before the _jparse repo_ is created, a new `jparse/test_jparse/test_JSON/` based on  the _party.json_ theme should be created.  The development of that new `jparse/test_jparse/test_JSON/` contents can wait until the time is ripe for the new _jparse repo_.

Perhaps what this enhancement needs is to:

- **MOVE** `jparse/test_jparse/test_JSON/` to `test_ioccc/test_JSON/`
- **MODIFY** the _test_ Makefile rules accordingly
- **MODIFY** `vermod.sh`  to default to `test_ioccc/test_JSON/`

**Comment by lcn2** [comment]
Work on this issue should start now. It needs be completed during the 0.9.9 pre-release phase as mentioned in [comment 1367866148](https://github.com/ioccc-src/mkiocccentry/issues/87#issuecomment-1367866148).

**Comment by xexyl** [comment]
This is so funny because I was thinking of this yesterday as I was falling asleep or getting ready to sleep - not sure which.

Please assign this to me.

**Comment by xexyl** [comment]
> ## Background
> > ...shouldn't `vermod.sh` be put under `jparse` and not `soup` ?
> 


> The propose of `vermod.sh` is _IOCCC specific_: to update version numbers that matter to the _mkiocccentry repo_.
> 

Yes I think I conflated in my head the verge and this script. I realised my error later but it's good I brought it up because this issue should be done with (and as I said earlier I thought of this too).

> While `jparse/test_jparse/test_JSON/` contains what seems to be _IOCCC data_, data that `vermod.sh` is designed to modify: what the _JSON parser_ needs are JSON test files. It just happens that the _IOCCC inspired_ `jparse/test_jparse/test_JSON/` happened to "_foot the bill_" (as that expression goes).
> 

Of course.

> ## move the test_JSON tree
> We suggest that `jparse/test_jparse/test_JSON/` should be moved to `test_ioccc/test_JSON/` and that `vermod.sh` be modified to impact that directory tree.
> 
> Now, the _JSON parser_ needs test JSON files of both valid and invalid JSON variety. So while the **CONTENT** under `jparse/test_jparse/test_JSON/` should be moved to `test_ioccc/test_JSON/`, the _JSON parser_ **STILL NEEDS** a `parse/test_jparse/test_JSON/` test tree: just something that is **NOT** _IOCCC / mkiocccentry_ related. We suggest that the "something" that the _JSON parser_ should do is to make the new `jparse/test_jparse/test_JSON/` based on the theme that `jparse/test_jparse/sample_JSON/party.json` uses.
> 
> For the above to be accomplished well, the respective tools that test the JSON parser need a clear mechanism to identify the directory under which the _test_JSON_ stuff is found. Moreover, the directory structure and filename structure under the current `jparse/test_jparse/test_JSON/` directory needs to be maintained.
> 
> The current `jparse/test_jparse/jparse_test.sh` does this with the `-d json_tree` argument that defaults to the "`$JSON_TREE`" (`./jparse/test_jparse/test_JSON`) value. So there may not be much that needs to change. Perhaps what is needed (in addition to the above melted tree move and new _test_JSON_ tree contents based on _party.json_ theme) is that the _test_ Makefile rule of `json/test_jparse/Makefile` needs to point at the _test_JSON_ tree contents based on the _party.json_ theme WHEREAS the _test_ Makefile rule of `test_ioccc/Makefile` would invoke `jparse/test_jparse/jparse_test.sh` executed with `-d test_ioccc/test_JSON` (containing the current JSON contents that is under `jparse/test_jparse/test_JSON/` now).
> 
I'm afraid the above is too much for me to parse right now :( too tired today .. was a difficult sleep.

> IMPORTANT NOTE: The new contents of `jparse/test_jparse/test_JSON/` are **NOT** needed right now. Before the _jparse repo_ is created, a new `jparse/test_jparse/test_JSON/` based on the _party.json_ theme should be created. The development of that new `jparse/test_jparse/test_JSON/` contents can wait until the time is ripe for the new _jparse repo_.
> 

The party theme is fun and I did start it as you know. But I get the feeling there should be more than just that. Do you have anything in mind? Well maybe we don't need to worry about it until the new repo is created. After all it's for that repo and we can add test files after it's been created.

> Perhaps what this enhancement needs is to:
> 
> * **MOVE** `jparse/test_jparse/test_JSON/` to `test_ioccc/test_JSON/`
> * **MODIFY** the _test_ Makefile rules accordingly
> * **MODIFY** `vermod.sh`  to default to `test_ioccc/test_JSON/`

I don't think this will take much effort but I'm tired enough where I think it would be unwise to try it so I'll delay it for now. I would imagine that I could do this in a sitting or two though.

**Comment by xexyl** [comment]
I believe this issue is entirely resolved. I'll do a pull request soon. I am afraid it's probably all I'll do today though. Not only am I very tired but I'm not feeling well either. Started yesterday. Not sick but feel physically off. I'll be fine.

**Comment by xexyl** [comment]
No .. there is a problem here. With the all ref rules. I'll have to leave soon for a while but I hope to address it after that. By 'soon' I mean now. I hope I can return to this in say an hour from now but it might be longer.

**Comment by xexyl** [comment]
Please see https://github.com/ioccc-src/mkiocccentry/pull/451/commits/13e9576010170826ee8e8b269bab5c90769d9fbb.

I think one thing can be changed to improve it .. I refer to the new `-s` option to `jparse_test.sh`. I noted it in the log. What do you think ?

I'll be away now and I don't expect to work on any other issues here today. In fact I mean not to but we'll see.

**Comment by lcn2** [comment]
> Please see https://github.com/ioccc-src/mkiocccentry/pull/451/commits/13e9576010170826ee8e8b269bab5c90769d9fbb.
> 
> 
> 
> This commit should resolve this issue. I think one thing can be changed to improve it .. I refer to the new `-s` option to `jparse_test.sh`. I noted it in the log. What do you think ?
> 
> 
> 
> I'll be away now and I don't expect to work on any other issues here today. In fact I mean not to but we'll see.

At first glance the `-s` looks like the best way to proceed.

**Comment by xexyl** [comment]
> > Please see [13e9576](https://github.com/ioccc-src/mkiocccentry/commit/13e9576010170826ee8e8b269bab5c90769d9fbb).
> > This commit should resolve this issue. I think one thing can be changed to improve it .. I refer to the new `-s` option to `jparse_test.sh`. I noted it in the log. What do you think ?
> > I'll be away now and I don't expect to work on any other issues here today. In fact I mean not to but we'll see.
> 
> At first glance the `-s` looks like the best way to proceed.

That's true. But maybe the default should be swapped so that jparse_test.sh by default checks the directory that the jparse files are in? That's what I meant. Though of course there might be another way to go about it entirely. This seemed logical though and easy to do.

**Comment by lcn2** [comment]
> > > Please see [13e9576](https://github.com/ioccc-src/mkiocccentry/commit/13e9576010170826ee8e8b269bab5c90769d9fbb).
> 
> > > This commit should resolve this issue. I think one thing can be changed to improve it .. I refer to the new `-s` option to `jparse_test.sh`. I noted it in the log. What do you think ?
> 
> > > I'll be away now and I don't expect to work on any other issues here today. In fact I mean not to but we'll see.
> 
> > 
> 
> > At first glance the `-s` looks like the best way to proceed.
> 
> 
> 
> That's true. But maybe the default should be swapped so that jparse_test.sh by default checks the directory that the jparse files are in? That's what I meant. Though of course there might be another way to go about it entirely. This seemed logical though and easy to do.

We agree on the default.  When the _jparse repo_ is created, that would be a good thing.  The _mkiocccentry repo_ can override the default by use of the `-s` option as you suggested.

**Comment by xexyl** [comment]
> > > > Please see [13e9576](https://github.com/ioccc-src/mkiocccentry/commit/13e9576010170826ee8e8b269bab5c90769d9fbb).
> > 
> > 
> > > > This commit should resolve this issue. I think one thing can be changed to improve it .. I refer to the new `-s` option to `jparse_test.sh`. I noted it in the log. What do you think ?
> > 
> > 
> > > > I'll be away now and I don't expect to work on any other issues here today. In fact I mean not to but we'll see.
> > 
> > 
> > > 
> > 
> > 
> > > At first glance the `-s` looks like the best way to proceed.
> > 
> > 
> > That's true. But maybe the default should be swapped so that jparse_test.sh by default checks the directory that the jparse files are in? That's what I meant. Though of course there might be another way to go about it entirely. This seemed logical though and easy to do.
> 
> We agree on the default. When the _jparse repo_ is created, that would be a good thing. The _mkiocccentry repo_ can override the default by use of the `-s` option as you suggested.

Please see https://github.com/ioccc-src/mkiocccentry/pull/452/commits/4ce7e0dd02089e2f17cdf745fd96c950ad3b048a.

Does this mean this issue is resolved entirely?

**Comment by lcn2** [comment]
> > > > > Please see [13e9576](https://github.com/ioccc-src/mkiocccentry/commit/13e9576010170826ee8e8b269bab5c90769d9fbb).
> > > 
> > > 
> > > > > This commit should resolve this issue. I think one thing can be changed to improve it .. I refer to the new `-s` option to `jparse_test.sh`. I noted it in the log. What do you think ?
> > > 
> > > 
> > > > > I'll be away now and I don't expect to work on any other issues here today. In fact I mean not to but we'll see.
> > > 
> > > 
> > > > 
> > > 
> > > 
> > > > At first glance the `-s` looks like the best way to proceed.
> > > 
> > > 
> > > That's true. But maybe the default should be swapped so that jparse_test.sh by default checks the directory that the jparse files are in? That's what I meant. Though of course there might be another way to go about it entirely. This seemed logical though and easy to do.
> > 
> > 
> > We agree on the default. When the _jparse repo_ is created, that would be a good thing. The _mkiocccentry repo_ can override the default by use of the `-s` option as you suggested.
> 
> Please see [4ce7e0d](https://github.com/ioccc-src/mkiocccentry/commit/4ce7e0dd02089e2f17cdf745fd96c950ad3b048a).
> 
> Does this mean this issue is resolved entirely?

It is close, but based on [comment 1368307696](https://github.com/ioccc-src/mkiocccentry/pull/452#issuecomment-1368307696). we'd say not completely.

**Comment by xexyl** [comment]
> > > > > > Please see [13e9576](https://github.com/ioccc-src/mkiocccentry/commit/13e9576010170826ee8e8b269bab5c90769d9fbb).
> > > > 
> > > > 
> > > > > > This commit should resolve this issue. I think one thing can be changed to improve it .. I refer to the new `-s` option to `jparse_test.sh`. I noted it in the log. What do you think ?
> > > > 
> > > > 
> > > > > > I'll be away now and I don't expect to work on any other issues here today. In fact I mean not to but we'll see.
> > > > 
> > > > 
> > > > > 
> > > > 
> > > > 
> > > > > At first glance the `-s` looks like the best way to proceed.
> > > > 
> > > > 
> > > > That's true. But maybe the default should be swapped so that jparse_test.sh by default checks the directory that the jparse files are in? That's what I meant. Though of course there might be another way to go about it entirely. This seemed logical though and easy to do.
> > > 
> > > 
> > > We agree on the default. When the _jparse repo_ is created, that would be a good thing. The _mkiocccentry repo_ can override the default by use of the `-s` option as you suggested.
> > 
> > 
> > Please see [4ce7e0d](https://github.com/ioccc-src/mkiocccentry/commit/4ce7e0dd02089e2f17cdf745fd96c950ad3b048a).
> > Does this mean this issue is resolved entirely?
> 
> It is close, but based on [comment 1368307696](https://github.com/ioccc-src/mkiocccentry/pull/452#issuecomment-1368307696). we'd say not completely, @xexyl .

Of course. I will fix it soon. Leaving now but when I am back I can fix it. Shouldn’t take more than a minute to fix and then I will test it too.

**Comment by xexyl** [comment]
Commit 8c43d694f69892e043cbd272569d21f53a8f9c6f should fix it.

**Comment by lcn2** [comment]
> Commit [8c43d69](https://github.com/ioccc-src/mkiocccentry/commit/8c43d694f69892e043cbd272569d21f53a8f9c6f) should fix it.

The problem with commit 8c43d694f69892e043cbd272569d21f53a8f9c6f  is that now `test_ioccc/ioccc_test.sh` only tests the files under `test_ioccc/test_JSON/general.json/`.

The following JSON test sets are **NOT** tested:

- jparse/test_jparse/test_JSON
- test_ioccc/test_JSON/info.json
- test_ioccc/test_JSON/auth.json

See commit 3e02cbcfeeb021f102ee0603c96167fc3032f193

Unfortunately, these test(s) failed:

```
    jparse/test_jparse/jparse_test.sh for info.json non-zero exit code: 1
    jparse/test_jparse/jparse_test.sh for auth.json non-zero exit code: 1
```

To close this issue, @xexyl, we suggest that you fix the broken tests, perhaps by fixing jparse_test.sh and/or by fix ioccc_test.sh.

**Comment by xexyl** [comment]
> > Commit [8c43d69](https://github.com/ioccc-src/mkiocccentry/commit/8c43d694f69892e043cbd272569d21f53a8f9c6f) should fix it.
> 
> The problem with commit [8c43d69](https://github.com/ioccc-src/mkiocccentry/commit/8c43d694f69892e043cbd272569d21f53a8f9c6f) is that now `test_ioccc/ioccc_test.sh` only tests the files under `test_ioccc/test_JSON/general.json/`.
> 
> The following JSON test sets were **NOT** tested:
> 
> * jparse/test_jparse/test_JSON
> * test_ioccc/test_JSON/info.json
> * test_ioccc/test_JSON/auth.json
> 

Strange ... I thought it did test it.

> See commit [3e02cbc](https://github.com/ioccc-src/mkiocccentry/commit/3e02cbcfeeb021f102ee0603c96167fc3032f193)
> 
> Unfortunately, these test(s) failed:
> 
> ```
>     jparse/test_jparse/jparse_test.sh for info.json non-zero exit code: 1
>     jparse/test_jparse/jparse_test.sh for auth.json non-zero exit code: 1
> ```
> 
> To close this issue, @xexyl, we suggest that you fix the broken tests, perhaps by fixing jparse_test.sh and/or by fix ioccc_test.sh.
> 
Sure. Though I might be too sleepy to fix it right yet - I just can't sleep either nor can I rest. I'm sitting up now with a cat between my legs but I might not be able to do anything here. Or maybe I will be able to (though I hope I can rest a bit soon).

> Sorry (tm Canada 🇨🇦) to leave `make test` in a broken state, but we must stop now .. too sleepy. Going to sleep now.

Hope you slept well! And better that you don't work on it when sleepy so it's fine. Besides you did a lot of work and it seems you got some of it resolved. I do have a concern but since it's to do with the man pages I'll make it a comment over there.

**Comment by xexyl** [comment]
> > > Commit [8c43d69](https://github.com/ioccc-src/mkiocccentry/commit/8c43d694f69892e043cbd272569d21f53a8f9c6f) should fix it.
> > 
> > 
> > The problem with commit [8c43d69](https://github.com/ioccc-src/mkiocccentry/commit/8c43d694f69892e043cbd272569d21f53a8f9c6f) is that now `test_ioccc/ioccc_test.sh` only tests the files under `test_ioccc/test_JSON/general.json/`.
> > The following JSON test sets were **NOT** tested:
> > 
> > * jparse/test_jparse/test_JSON
> > * test_ioccc/test_JSON/info.json
> > * test_ioccc/test_JSON/auth.json
> 
> Strange ... I thought it did test it.
> 
> > See commit [3e02cbc](https://github.com/ioccc-src/mkiocccentry/commit/3e02cbcfeeb021f102ee0603c96167fc3032f193)
> > Unfortunately, these test(s) failed:
> > ```
> >     jparse/test_jparse/jparse_test.sh for info.json non-zero exit code: 1
> >     jparse/test_jparse/jparse_test.sh for auth.json non-zero exit code: 1
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > To close this issue, @xexyl, we suggest that you fix the broken tests, perhaps by fixing jparse_test.sh and/or by fix ioccc_test.sh.
> 
> Sure. Though I might be too sleepy to fix it right yet - I just can't sleep either nor can I rest. I'm sitting up now with a cat between my legs but I might not be able to do anything here. Or maybe I will be able to (though I hope I can rest a bit soon).
> 
> > Sorry (tm Canada 🇨🇦) to leave `make test` in a broken state, but we must stop now .. too sleepy. Going to sleep now.
> 
> Hope you slept well! And better that you don't work on it when sleepy so it's fine. Besides you did a lot of work and it seems you got some of it resolved. I do have a concern but since it's to do with the man pages I'll make it a comment over there.


Actually looking at it it seems many many files failed and I'm not sure why. This is very strange since the files didn't fail before. For instance:

```sh
$ jparse/test_jparse/jparse_test.sh -J 0 -d ./test_ioccc/test_JSON -s auth.json -j jparse/jparse jparse/test_jparse/json_teststr.txt 2>&1|grep -c FILE:
213
$ jparse/test_jparse/jparse_test.sh -J 0 -d ./test_ioccc/test_JSON -s info.json -j jparse/jparse jparse/test_jparse/json_teststr.txt 2>&1|grep -c FILE:
216
```

Why? The only thing I can think of is that the test is inverted so that it thinks they should pass? I'll dig deeper into this after I address other comments - if I can at least. If not I'll do it later on.

**Comment by xexyl** [comment]
See updated comment https://github.com/ioccc-src/mkiocccentry/issues/449#issuecomment-1368949387 for the fix.

**Comment by lcn2** [comment]
> See updated comment [#449 (comment)](https://github.com/ioccc-src/mkiocccentry/issues/449#issuecomment-1368949387) for the fix (or else: [d37d15d](https://github.com/ioccc-src/mkiocccentry/commit/d37d15dd4dcf4d57c6e5f193921d41744463bd1a)).
> 
> Remember the general.json/bad files have to be invalid json files and jparse_test.sh DOES test those. The info and auth files are valid JSON but the bad subdirectory under each are bad according to chkentry. However if say info.json/good is not valid json it'll fail for obvious reasons as it means the parser failed (validly so). If the bad/ files fail it means that the parser couldn't parse it OR the file is valid according to the chkentry tool and in either case it is a fail.
> 
> Thus I believe this issue is now resolved. Am I right?

Correct, and thank you @xexyl

**Comment by xexyl** [comment]
> > See updated comment [#449 (comment)](https://github.com/ioccc-src/mkiocccentry/issues/449#issuecomment-1368949387) for the fix (or else: [d37d15d](https://github.com/ioccc-src/mkiocccentry/commit/d37d15dd4dcf4d57c6e5f193921d41744463bd1a)).
> > Remember the general.json/bad files have to be invalid json files and jparse_test.sh DOES test those. The info and auth files are valid JSON but the bad subdirectory under each are bad according to chkentry. However if say info.json/good is not valid json it'll fail for obvious reasons as it means the parser failed (validly so). If the bad/ files fail it means that the parser couldn't parse it OR the file is valid according to the chkentry tool and in either case it is a fail.
> > Thus I believe this issue is now resolved. Am I right?
> 
> Correct, and thank you @xexyl

Happy to help my friend!
